components/ActionSheet/src/MDCActionSheetController.m

```diff
@@ -0,0 +1,343 @@
+                      handler:(void (^__nullable)(MDCActionSheetAction *action))handler {
+  self = [super init];
+  if (self) {
+    _title = [title copy];
+    _image = [image copy];
+    _completionHandler = [handler copy];
+  }
+  return self;
+}
+
+- (id)copyWithZone:(__unused NSZone *)zone {
+  MDCActionSheetAction *action = [[self class] actionWithTitle:self.title
+                                                         image:self.image
+                                                       handler:self.completionHandler];
+  return action;
+}
+
+@end
+
+@interface MDCActionSheetController () <MDCBottomSheetPresentationControllerDelegate,
+    UITableViewDelegate, UITableViewDataSource>
+
+@end
+
+@implementation MDCActionSheetController {
+  MDCActionSheetHeaderView *_header;
+  UITableView *_tableView;
+  MDCBottomSheetTransitionController *_transitionController;
+  NSMutableArray<MDCActionSheetAction *> *_actions;
+}
+
+@synthesize mdc_adjustsFontForContentSizeCategory = _mdc_adjustsFontForContentSizeCategory;
+
++ (instancetype)actionSheetControllerWithTitle:(NSString *)title message:(NSString *)message {
+  return [[MDCActionSheetController alloc] initWithTitle:title message:message];
+}
+
++ (instancetype)actionSheetControllerWithTitle:(NSString *)title {
+  return [MDCActionSheetController actionSheetControllerWithTitle:title message:nil];
+}
+
+- (instancetype)init {
+  return [MDCActionSheetController actionSheetControllerWithTitle:nil message:nil];
+}
+
+- (instancetype)initWithTitle:(NSString *)title message:(NSString *)message {
+  self = [super initWithNibName:nil bundle:nil];
+  if (self) {
+    _actions = [[NSMutableArray alloc] init];
+    _transitionController = [[MDCBottomSheetTransitionController alloc] init];
+    _transitionController.dismissOnBackgroundTap = YES;
+    /**
+     "We must call super because we've made the setters on this class unavailable and overridden
+     their implementations to throw assertions."
+     */
+    super.transitioningDelegate = _transitionController;
+    super.modalPresentationStyle = UIModalPresentationCustom;
+    _tableView = [[UITableView alloc] initWithFrame:CGRectZero style:UITableViewStylePlain];
+    _transitionController.trackingScrollView = _tableView;
+    _tableView.autoresizingMask = (UIViewAutoresizingFlexibleWidth
+                                   | UIViewAutoresizingFlexibleHeight);
+    _tableView.delegate = self;
+    _tableView.dataSource = self;
+    _tableView.estimatedRowHeight = 56.f;
+    _tableView.rowHeight = UITableViewAutomaticDimension;
+    _tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
+    [_tableView registerClass:[MDCActionSheetItemTableViewCell class]
+       forCellReuseIdentifier:ReuseIdentifier];
+
+    _header = [[MDCActionSheetHeaderView alloc] initWithFrame:CGRectZero];
+    _header.title = [title copy];
+    _header.message = [message copy];
+    self.backgroundColor = [UIColor whiteColor];
+  }
+
+  return self;
+}
+
```

Comment:
Because we still support iOS 8 we need to manually remove and notification handlers that we may have added. In our case this would be the Dynamic Type handler.  iOS 9 handles this for us, so this is a temporary thing.

- (void)dealloc {
  //TODO(#2651): Remove once we drop iOS8 support
  [[NSNotificationCenter defaultCenter] removeObserver:self];
}